add fs backend

cosmicexplorer opened the pull request:
Hello! This is an awesome project!! ^_^

### Problem
In pantsbuild/pants#9973, I tried to create a standalone binary wrapping the `cargo-fetcher` functionality that the pants build tool would invoke, but realized it would be far easier for pants to consume if I just used the filesystem to cache data, because pants creates a separate isolated chroot for invoking each subprocess.

Additionally, the S3 backend is very difficult to test without hitting frequent flakes, and is currently disabled. We would like to be able to test all of our logic without being dependent on some flaky network resource.

### Solution
This PR adds an `fs` backend that wraps a directory in the filesystem, and uses it to turn back on some testing that used to be flaky!

#### `FilesystemDB`
The proposed filesystem backend is implemented in a new file `fs.rs`. The caching logic implemented by `FilesystemDB` (a key-value store) is:
1. serialize the key to string as a json object,
2. then use a `Fingerprint` of the bytes of that json object as a filename.

`FilesystemDB::insert()` will do something similar with the value, serializing it to string as a json object, then writing the bytes of that json string to the filename specified by the `Fingerprint` key.

This produces a flat directory with files named after the `sha256sum`s of their inputs. Each `FilesystemDB` should be constructed with its own such directory.

#### `FSBackend`
The new `FSBackend` uses several of these `FilesystemDB` instances to implement the different `Backend` operations.

### Result
When executing the `cargo-fetcher` binary on the command line, if the argument to `--url` has a `file://` scheme, the local path that the `file://` url points to will be used as a flat content-addressed database to satisfy the `Backend` interface.

cosmicexplorer commented on Cargo.toml:
```diff
@@ -43,14 +48,15 @@ tracing-futures = "0.2.2"
 url = "2.1.1"
 walkdir = "2.3.1"
-zstd = "0.5.1+zstd.1.4.4"
+zstd = "0.5.1"
```
This removes a warning message that says the zstd version is already fully specified.

cosmicexplorer commented:
I'd love to know what kind of testing this project typically requires for new features like this! I have lots of patience for adding great tests!

cosmicexplorer commented:
Also, I'd like some help perhaps to debug why, when `--features s3` is enabled, there's this error  compiling the `futures` crate complaining about a macro from `future_utils::...` not existing. The failure reproduces locally as well, and I've only gotten past it by editing the erroring source file in the cache directory to remove that excerpt!

Jake-Shadle commented:
Sorry, it looks some errors have crept in since the last time I updated this crate, I've updated the dependencies and fixed up our deny configuration, so rebasing this PR on master should resolve most (all?) of the issues that are currently blocking this PR.

As far as tests go, I actually wouldn't mind if you replaced the current usages of the S3 backend to use the FS one, as we had to disable most of the tests due to how flaky minio was. I doubt all tests will "just work" since I was doing some experiments before I left in, but just keeping those failing tests as `#[ignore]` and using the FS backend for the rest of them would be great actually.

cosmicexplorer commented:
Oh! I've figured it out! The `futures` crate version `0.3.4` has allowed an impossible feature configuration to slip through! Adding an explicit dependency to `futures-util = { version = "0.3.4", features = ["async-await-macro"] }` fixes it -- I'll push a commit.

cosmicexplorer commented:
> Sorry, it looks some errors have crept in since the last time I updated this crate, I've updated the dependencies and fixed up our deny configuration, so rebasing this PR on master should resolve most (all?) of the issues that are currently blocking this PR.

Thanks so much! I will rebase now!

> As far as tests go, I actually wouldn't mind if you replaced the current usages of the S3 backend to use the FS one, as we had to disable most of the tests due to how flaky minio was. I doubt all tests will "just work" since I was doing some experiments before I left in, but just keeping those failing tests as #[ignore] and using the FS backend for the rest of them would be great actually.

This sounds great! I can do this!

cosmicexplorer commented:
@Jake-Shadle done, in a second commit! This one should pass lint too!

cosmicexplorer commented:
Passed CI!
